### DISCUSSION TITLE
Welcome to background_downloader Discussions!

### DISCUSSION THREAD

**Discussion by 781flyingdutchman** [781flyingdutchman/background_downloader#146] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

 If you have an issue with background_downloader then please use the _Issues_ tab, so we can keep this _Discussions_ tab about ideas etc.

**Comment by Darryl1702** [comment]
How to delete downloaded file(s)?

**Reply by 781flyingdutchman** [reply]
Just like you would delete any file in Dart:
```
final path = await task.filePath();
await File(path).delete();
```

**Comment by albertznk** [comment]
Hello can someone help me when i want to track the status using  this code in initState. 
 
FileDownloader().updates.listen((update) {
      switch (update) {
        case TaskStatusUpdate _:
          // process the TaskStatusUpdate, e.g.
        switch (update.status) {
          case TaskStatus.complete:
          print('Task ${update.task.taskId} success!');
          
          case TaskStatus.canceled:
          print('Download was canceled');
          
          case TaskStatus.paused:
          print('Download was paused');
          
          default:
          print('Download not successful');
          }

        case TaskProgressUpdate _:
          // process the TaskProgressUpdate, e.g.
          progressUpdateStream.add(update); // pass on to widget for indicator
      }
    });

And i got the error 
"The following StateError was thrown building Builder:
Bad state: Stream has already been listened to."   Can someone suggest how i can fix i try to use stream like    StreamController<TaskProgressUpdate> progressUpdateStream = StreamController()..stream.asBroadcastStream();  and it's still got same error.

**Comment by 781flyingdutchman** [comment]
Hi @zarnikyaw,  the design of the downloader is such that you should start listening to `updates` as soon as possible in your app's life, and to keep listening. The error you are getting says that you try to listen to `updates` more than once, and this is typically caused by you placing the `.listen` logic in a piece of code that can be executed more than once, such as the `initState` of a Widget.  If another screen covers the original screen and then is dimissed, the `initState` of the Widget may be called again, re-executing the code and leading to this error.  That tells you there is a problem with how you have architected your application (w.r.t. listening to updates from the downloader), as you should never attempt to listen to updates more than once.

Generally, the best way to implement this 'listen once and always' strategy is to create a [Singleton](https://flutterbyexample.com/lesson/singletons) object (or a [BLOC](https://pub.dev/packages/flutter_bloc) if you use that pattern) when your app starts up, and use that object to hold state for all things that relate to downloads. During the initialization of the singleton object (which will by definition only happen once) you define your `.listen` logic, and for example update the state of the downloads in progress based on the `updates` messages.  In your app, you then refer to the singleton object whenever you want to understand the state of your downloads. Often, it makes sense to also manage the interactions with the `FileDownloader` in that same singleton, so that it 'wraps' the `FileDownloader` and the singleton becomes your interface to all things related to downloads. That kind of separation has the added benefit that if you wanted to change the download mechanism (e.g. use a different download plugin) you only need to update your singleton, and the interface to your app does not change.

**Reply by albertznk** [reply]
Thanks.
